File: python/epi/levenshtein_distance.py
#!/usr/bin/env python3
#
# Description:
#
# Find the Levenshtein distance between two given words. The Levenshtein
# distance is defined as the minimum number of edits needed to morph one word
# to another word, where the edit operations comprise of 1) replacing a letter,
# 2) inserting a letter, and 3) deleting a letter.

import sys

def find_levenshtein_distance(s1, s2):
    memo = [[-1 for _ in s2] for _ in s1]
    for i in range(len(s1)):
        for j in range(len(s2)):
            insertOp = memo[i - 1][j] if i - 1 >= 0 else None
            deleteOp = memo[i][j - 1] if j - 1 >= 0 else None
            replaceOp = memo[i - 1][j - 1] if i - 1 >= 0 and j - 1 >= 0 else None
            if insertOp == None and deleteOp == None and replaceOp == None:
                minOp = 0
            else:
                minOp = min(insertOp if insertOp != None else sys.maxsize,
                            deleteOp if deleteOp != None else sys.maxsize,
                            replaceOp if replaceOp != None else sys.maxsize)
            memo[i][j] = 0 if s1[i] == s2[j] else 1
            memo[i][j] += minOp
    return memo[-1][-1]

File: java/epi/src/main/java/moe/ijnji/epi/LevenshteinDistance.java
// Description:
//
// Find the Levenshtein distance between two given words. The Levenshtein 
// distance is defined as the minimum number of edits needed to morph one word
// to another word, where the edit operations comprise of 1) replacing a letter,
// 2) inserting a letter, and 3) deleting a letter.

package moe.ijnji.epi;

import java.lang.Math;
import java.util.Arrays;
import java.util.List;


public class LevenshteinDistance {

  private static int levenshteinRecurse(String s1, String s2) {
    if (s1.equals("")) return s2.length();
    if (s2.equals("")) return s1.length();
    if (s1.charAt(0) == s2.charAt(0)) {
      return levenshteinDistance(s1.substring(1), s2.substring(1));
    } else {
      int deleteOp = levenshteinDistance(s1.substring(1), s2);
      int insertOp = levenshteinRecurse(s1, s2.substring(1));
      int replaceOp = levenshteinRecurse(s1.substring(1), s2.substring(1));
      return 1 + Math.min(deleteOp, Math.min(insertOp, replaceOp));
    }
  }

  public static int levenshteinDistance(String s1, String s2) {
    return levenshteinRecurse(s1, s2);
  }

  private static void smallTest() {
    List<String>  words;
    words = Arrays.asList("", "");
    //expect(levenshteinDistance(words.get(0), words.get(1))).toBe(0);
    words = Arrays.asList("a", "a");
    //expect(levenshteinDistance(words.get(0), words.get(1))).toBe(0);
    words = Arrays.asList("z", "x");
    //expect(levenshteinDistance(words.get(0), words.get(1))).toBe(1);
    words = Arrays.asList("Friday", "Friday");
    //expect(levenshteinDistance(words.get(0), words.get(1))).toBe(0);
    words = Arrays.asList("Saturday", "Sundays");
    //expect(levenshteinDistance(words.get(0), words.get(1))).toBe(4);
    words = Arrays.asList("Carthorse", "Orchestra");
    //expect(levenshteinDistance(words.get(0), words.get(1))).toBe(8);
  }

  public static void main(String[] args) {
    smallTest();
  }

}